In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

34 67
101

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

 5 6 
11


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


$$n^2 - n \ \ \text{Prorducts}, n^2 \ \ i \neq j \ \text{Comparisons}= O(n^2)$$

In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

5
1
2
3
4
5
20


### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product < input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [ ]:
200000*200000

40000000000

In [ ]:
!./max_pairwise_product < input2.txt

1345294336


In [ ]:
%%writefile output2.txt
40000000000

Overwriting output2.txt


In [ ]:
!./max_pairwise_product < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt
!./max_pairwise_product_long < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 3  stress tests 3 testing cases files input3.txt/output3.txt .. input5.txt/output5.txt

---



### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end)

In [ ]:
%%writefile Test4Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <199999; i++){
     MyFile << i <<" ";
 }
 MyFile << 200000 << " ";
 MyFile << 200000 ;
  // Close the file
  MyFile.close();
}

Writing Test4Write.cpp


In [ ]:
!g++ Test4Write.cpp -o Test4Write
!./Test4Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

40000000000

real	2m22.769s
user	2m21.107s
sys	0m0.080s


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long m = 0LL;                                                       //1

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n
            execints += 2;

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            execints++;

            execints++;
            if ( max_product < aux_product)  {                               //(n-1)n/2

				        max_product = aux_product;                                   //[0, (n-1)n/2]
                execints++;
             }
        }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
}

int main() {                                                 //N inst exec by line executed
    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }

    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

289
116
80003472079


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077)

625028688.1015625

In [ ]:
# Run 10 times with the worst case input in C++ use the maximum time and calculate the maximun n that can be solved in 1 second

In [ ]:
# Verify

## Python

In [32]:
def max_pairwise_product_inn(numbers):
    n = len(numbers)
    max_product = 0

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product,
                numbers[first] * numbers[second])

    return max_product

In [33]:
%%writefile max_pair.py

if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product_inn(input_numbers))

Overwriting max_pair.py


In [10]:
import random
import math

def gen_stress_test(filename, n):
    file = open(filename, "w")
    file.write(f'{n}\n')

    for i in range(n):
        file.write(f' {i}')

    file.close()


if __name__ == '__main__':
    k = 6
    for i in range(1, k + 1):
        n = math.ceil(random.random() * (200_000 - 50_000) + 50_000)
        gen_stress_test(f'input_{i}.txt', n)

In [ ]:
# Run 10 times with the worst case input in Python use the maximum time and calculate the maximun n that can be solved in 1 second
#< input_1.txt > output_1.txt

!python3 max_pair.py < input_1.txt > output_1.txt
!python3 max_pair.py < input_2.txt > output_2.txt
!python3 max_pair.py < input_3.txt > output_3.txt
!python3 max_pair.py < input_4.txt > output_4.txt
!python3 max_pair.py < input_5.txt > output_5.txt
!python3 max_pair.py < input_6.txt > output_6.txt


^C


In [ ]:
# Verify

!time python3 max_pair.py < input_4.txt

3328309172

real	7m55.031s
user	7m49.454s
sys	0m0.162s


In [25]:
class TimedOp():
    def __init__(self, n, total_time, get_instructions):
        self.n = n
        self.total_time = total_time
        self.get_instructions = get_instructions
        self.inst_per_total =  self.get_instructions(self.n)

    def time_per_inst(self):
        return self.total_time / self.get_instructions(self.n)

    def insts_per_time(self, time):
        return time / (self.total_time / self.inst_per_total)

In [26]:
n = 178034
total_time = 449.4
get_instructions = lambda n: 9 + (7 * n)  + 2.5 * ((n - 1) * n)

inneficient = TimedOp(n, total_time, get_instructions)
print(inneficient.inst_per_total)
print(inneficient.time_per_inst())
print(inneficient.insts_per_time(1))

79241064052.0
5.671301936393639e-09
176326355.25589675


In [12]:
n_for_sec = 8399

In [27]:
gen_stress_test("input_1sec.txt", n_for_sec)

In [28]:
!time python3 max_pair.py < input_1sec.txt

70518006

real	0m5.109s
user	0m5.071s
sys	0m0.012s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{Comparisons}$$

### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [39]:
def max_pairwise_product(nums):
    nums.sort()
    return nums[0] * nums[1]

In [20]:
%%writefile max_sort.py

if __name__ == '__main__':
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product(input_numbers))

Writing max_sort.py


In [21]:
gen_stress_test("input_for_sort.txt", 10_000_000)

Test with inputs 1 and 2



In [ ]:
!time python3 max_sort.py < input_for_sort.txt

0

real	0m5.477s
user	0m3.526s
sys	0m1.818s


Test and time it with input 3

In [ ]:
sorting = TimedOp(10_000_000, 5.477, lambda x: x * math.log(x))

print(sorting.max_n_for(1))

0.5477000000000001


20 stress tests

In [ ]:
gen_stress_test("input_1sec_sorted.txt", 547000)

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
!time python3 max_sort.py < input_1sec_sorted.txt

0

real	0m0.238s
user	0m0.186s
sys	0m0.046s


Calculate estimated time by instruction and number of instructions by second

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [34]:
%%writefile max_efficient.py

def max_pairwise_efficient():
    n = int(input())
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for i in range(n):
        new = int(input())
        temp = biggest_1
        if new > biggest_1: biggest_1 = new
        if temp > biggest_2: biggest_2 = temp

    return biggest_1 * biggest_2


if __name__ == '__main__':
    print(max_pairwise_efficient())

Writing max_efficient.py


Test with inputs 1 and 2




In [35]:
!time python3 max_efficient.py < input_1.txt

Traceback (most recent call last):
  File "/content/max_efficient.py", line 17, in <module>
    print(max_pairwise_efficient())
  File "/content/max_efficient.py", line 8, in max_pairwise_efficient
    new = int(input())
ValueError: invalid literal for int() with base 10: ' 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 6

real	0m0.038s
user	0m0.032s
sys	0m0.002s


Automated stress test for multiple methods.

In [42]:
class StressTest:
    def __init__(self, a, b, k = 100_000_000_000, max = 10_000, min = 2):
        self.k = k
        self.max = max
        self.min = min
        self.a = a
        self.b = b

    def run(self):
        for i in range(1, self.k + 1):
            n = math.ceil(random.random() * (self.max - self.min) + self.min)
            inp = [int(x) for x in range(1, n)]
            res_a = self.a(inp)
            res_b = self.b(inp)

            if res_a != res_b:
                raise ValueError(f"both methods didn't match for {n=} {res_a=}, {res_b=}")


In [43]:
st = StressTest(max_pairwise_product_inn, max_pairwise_product)
st.run()

ValueError: ignored